polus-vector-to-label-plugin/src/main.py

```diff
@@ -0,0 +1,186 @@
+    logger.info('inpDir = {}'.format(inpDir))
+    outDir = args.outDir
+    logger.info('outDir = {}'.format(outDir))
+    cellprob_threshold = args.cellprobThreshold
+    flow_threshold= args.flowThreshold
+    stitch_threshold=args.stitchThreshold
+    rescale = np.ones(1)
+    niter = 1 / rescale[0] * 200
+    # Surround with try/finally for proper error catching
+    try:
+        logger.info('Initializing ...')
+        # Get all file names in inpDir image collection
+        root = zarr.open(str(Path(inpDir).joinpath('flow.zarr')),mode='r')
+        count=0
+
+        # Loop through files in inpDir image collection and process
+        for file_name, vec in root.groups():
+            logger.info(
+                'Processing image ({}/{}): {}'.format(count + 1, len([file_name for file_name, _ in root.groups()]),
+                                                      file_name))
+
+            metadata = vec.attrs['metadata']
+
+
+            tile_size = min(1024, root[file_name]['vector'].shape[1])
+            lblcnt_max = 0
+
+            path = Path(outDir).joinpath(str(file_name))
+            xml_metadata = OmeXml.OMEXML(metadata)
+
+            new_img = -1
+            with  BioWriter(file_path=Path(path), backend='python', metadata=xml_metadata) as bw:
+                bw.dtype=np.dtype(np.uint32)
+                for x in range(0, root[file_name]['vector'].shape[1], tile_size):
+                    x_max = min([root[file_name]['vector'].shape[1], x + tile_size])
+                    for y in range(0, root[file_name]['vector'].shape[1], tile_size):
+                        y_max = min([root[file_name]['vector'].shape[1], y + tile_size])
+                        # Iterating over Z dimension
+                        for z in range(0,root[file_name]['vector'].shape[2],1):
+                            tile = root[file_name]['vector'][y:y_max, x:x_max, z:z + 1, :, :]
+
+                            tile=tile.transpose((2, 0, 1, 3, 4)).squeeze()
+                            logger.info(
+                                'Calculating flows and masks  for tile [{}:{},{}:{},{}:{}]'.format(y, y_max, x,
+                                                                                                    x_max, z, z + 1))
+                            cellprob = tile[..., -1]
+                            dP = np.stack((tile[..., 0], tile[..., 1]), axis=0)
+                        # Computing flows for the tile
+                            p = mask.follow_flows(-1 * dP * (cellprob > cellprob_threshold) / 5.,
+                                              niter=niter, interp=True)
+                        # Generating masks for the tile
+                            maski = mask.compute_masks(p, cellprob, dP, new_img,z, cellprob_threshold,
+                                                            flow_threshold)
+
+                            new_tile=maski
+                            if z > 0 and stitch_threshold > 0 :
```

Comment:
The way you are doing tiling here is going to be problematic for large images. The z-values should be in the outermost loop. This should not work properly for 3d data, because the tiled tiff format requires that each plane be written for a specific z-slice before the next z-slice can be written. It appears as though you are trying to write two z-slices at a time, which will lead to either an error or an improperly written file.